File: code/ch13/13.4.2.best_time_to_buy_and_sell_stock_with_cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() <= 1)
            return 0;
        
        int b0 = -prices[0];
        int b1 = b0;
        int s0 = 0, s1 = 0, s2 = 0;
        
        for (int i = 0; i < prices.size(); i++) {
            b0 = max(b1, s2 - prices[i]);
            s0 = max(s1, b0 + prices[i]);
            s2 = s1;
            s1 = s0;
            b1 = b0;
        }
        
        return s0;
    }
};

File: code/ch13/13.4.2.best_time_to_buy_and_sell_stock_with_cooldown.py
from typing import List


class Solution:

    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1:
            return 0
        b0 = -prices[0]
        b1 = b0
        s0, s1, s2 = 0, 0, 0
        for i in range(0, len(prices)):
            b0 = max(b1, s2 - prices[i])
            s0 = max(s1, b0 + prices[i])
            s2 = s1
            s1 = s0
            b1 = b0
        return s0
